[WIP] First stab at source tree guide in docs/ directory.

jeff-cohere opened the pull request:
Closes #394

jeff-cohere commented:
Here's the source code guide. The other thing Luca wanted was to replace `components/scream/README.md` with a simple description of SCREAM with links to documentation, and to move the build/testing information into `docs/` somewhere. Should we call that `docs/building.md` or something?

bartgol commented on components/scream/docs/source-tree.md:
```diff
@@ -0,0 +1,63 @@
++ `dynamics`: Here's where HOMME lives within SCREAM, along with code for
+  interfacing with it using SCREAM's data structures.
++ `interface`: Glue code for embedding SCREAM within E3SM as an atmosphere
+  process.
```
"process" -> "component". The word process is already used as a piece of the atmosphere.

jeff-cohere commented:
Thanks, Luca. I've added `docs/build.md`. The Kokkos stuff seemed out of date, so I updated the build instructions. Let me know if I got something wrong.

The remaining step is to discard the current contents of `components/scream/README.md` and replace them with a brief description of the project and some links. @PeterCaldwell , do you have any thoughts about what specifically goes in there, or should I just whip something up?

PeterCaldwell commented on components/scream/docs/source-tree.md:
```diff
@@ -0,0 +1,63 @@
+
+In addition, you'll notice the following files in `components/scream`:
+
++ `CMakeLists.txt`: The CMake file that defines SCREAM's configuration/build
```
Since CMakeLists.txt and tests/ is included in every directory in components/scream/ (or at least in src/ subdirectories), perhaps it would be worth moving mention of these to a special section about patterns found in all directories?

bartgol replied:
I think we could actually avoid mentioning the CMakeLists.txt files. If one doesn't know cmake at all, we have a bigger problem than the documentation. Honestly, we could probably skip the description of the individual files, and focus only on the directories. That's what the user need to know, in order to find the code he/she needs (or in order to find out where he/she should put his/her code).

jeff-cohere replied:
Up to you. I just mentioned them to give the reader the impression that these things are around and that the build system isn't magical.

I don't think that many people "know" CMake, to be honest. And the CMake docs are pretty bad for beginners. :-)

bartgol replied:
Fair point. Then we could simply say that files in the repo that start with `CMake` or `CTest` or have extension `.cmake` are files that contain cmake-syntax code, used to configure, build, and test the project. And maybe add a link to some cmake documentation (even if not great) for those who know little-to-nothing about cmake.

PeterCaldwell replied:
Yeah, a lot of people who know nothing about CMake - or C++ - will end up cutting their teeth on our code base... so yeah, we have bigger problems, but every little bit helps.

PeterCaldwell commented on components/scream/docs/source-tree.md:
```diff
@@ -0,0 +1,63 @@
+
+Herein lіes the source code for SCREAM. Broadly, here's where things are:
+
++ `control`: Contains the atmosphere driver and basic tests for it.
```
@bartgol and @AaronDonahue - would 'driver' be a better name for the directory containing the AD? I don't think we need to replicate CAM wording if it doesn't make sense. This isn't something we should (?) do in this PR...

bartgol replied:
I think I picked 'control' cause that would allow me to type `c<TAB>` and get autocompletion, while `d<TAB>` didn't, because of the presence of the dynamics folder as well. And having to type `dy<TAB>` to get in the dynamics folder was annoying, since the 'y' key is one of the most unconvenient to type.

That's incidentally the same reason why I picked 'interface' rather than 'coupling' or 'coupler' for the code related to the component coupler. It would clash with 'control'.

#lazy

PeterCaldwell replied:
not sure how to respond to that. "Laziness is a virtue"? Let's leave things as-is for now. I think having similar objects with different names will make onboarding harder for domain scientists looking at the code for the first time, so I may still want to change things later.

PeterCaldwell commented on components/scream/docs/source-tree.md:
```diff
@@ -0,0 +1,63 @@
+  + `p3`: The C++/Kokkos implementation of P3 microphysics within SCREAM.
+  + `shoc`: The C++/Kokkos implementation of SHOC macrophysics within SCREAM.
+  + `rrtmgp`: A stub for the radiation processes as represented in SCREAM.
+  + `common`: Utilities and data structures common to these processes.
```
I still want to rename 'common' -> 'share' to match terminology used elsewhere in our directory structure. Again, not for this PR.

bartgol replied:
I'm not sure how we landed on that name, but we might have picked common precisely to avoid confusion with the other (main) share folder. Personally, I do think 'common' (or 'generic') is better than share, in this context. But I don't have a very strong opinion about it.

PeterCaldwell replied:
Again, let's leave things as-is and see if divine inspiration strikes us in the next few weeks about what to do.

PeterCaldwell commented on components/scream/docs/source-tree.md:
```diff
@@ -0,0 +1,63 @@
+
+You'll also see some files in the `src/` directory itself:
+
++ `CMakeLists.txt`: The build definition file that points CMake at the various
```
I think you can delete this line since you mention above that CMakeLists.txt is in every directory.

PeterCaldwell commented:
Regarding components/scream/README.md: 

1. Let's link to the SCREAM confluence space https://acme-climate.atlassian.net/wiki/spaces/NGDNA/overview and mention in particular the Documentation tab. Maybe also say that only members of E3SM have access to these pages.
2. Mention that F90 versions of SCREAM code can be found in components/cam/src/ even though they are used by components/scream/ for C++ vs F90 BFB testing. Perhaps also mention that the F90 version of SCREAM uses the CAM/E3SMv1 process coupling methodology, while C++ SCREAM uses the AD.
3. Note that build.md describes how to run SCREAM stand alone unit tests.

bartgol commented on components/scream/docs/build.md:
```diff
@@ -0,0 +1,93 @@
+paths:
+
++ `${RUN_ROOT_DIR}`: the root directory where SCREAM is built and run
++ `${SCREAM_SRC_DIR}`: the directory into which you've cloned the `scream` repo
```
Is this the repo root dir, or is it components/scream? I think you mean the latter, but your description makes it sound like the git repo root dir.

jeff-cohere replied:
It's supposed to be the repo root directory. I think I've fixed references to it.

PeterCaldwell commented on components/scream/docs/build.md:
```diff
@@ -0,0 +1,93 @@
+# Building and Testing Standalone SCREAM
+
```
Thanks for moving this stuff from README.md to build.md. Makes much more sense. I think "standalone SCREAM" isn't quite the right word since this description allows us to test SCREAM unit tests...

jeff-cohere replied:
Fixed.

bartgol commented on components/scream/docs/build.md:
```diff
@@ -0,0 +1,93 @@
+
+Here, we've configured a `Debug` build to make it easier to find and fix errors.
+For performance testing, you should configure a `Release` build and make use of
+other options, depending on your architecture.
```
Should we add a list of possible options? Like `SCREAM_PACK_SIZE` or `SCREAM_FPMODEL` or `SCREAM_FPE`? I don't want to go too much overboard, but we do use some `SCREAM_blah` option in pretty much every build.

jeff-cohere replied:
I like that idea, but maybe that requires more people to chime in. Maybe we can start with this and iterate.

bartgol replied:
I can help.

PeterCaldwell replied:
I like the idea of including options. This build.md file is somewhat redundant with https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/1264386127/Running+SCREAM+Tests though, which makes me nervous. Perhaps provide a link to this html page in build.md? OR maybe we should only document how to run in confluence (in which case people outside e3sm would never have access to info on how to run?

jeff-cohere replied:
I think the closer to the code the information lives, the more likely it is to be kept up to date. :-) That and your last comment make me think that it's better to have this information in the repo. But one thing we can do is see where most of the questions originate. :-)

bartgol replied:
I think they are both important. Yes, if you have access to both, there's redundancy. And yes, it is harder to keep two documentations up to date. However, confluence has restricted access, so I'd like to keep some documentation in the repo (which is public). Not that we are very interested in users support outside of E3SM ecosystem, but it feels "right". Besides, Jeff is right, that documentation should be as close as possible to the code. I think a new user is more likely to browse the doc folder than navigating the (big) confluence space.

bartgol commented on components/scream/docs/build.md:
````diff
@@ -0,0 +1,93 @@
+```
+mkdir -p $RUN_ROOT_DIR/test
+cmake \
+    -D CMAKE_INSTALL_PREFIX=${RUN_ROOT_DIR}/install \
````
Do we support installation? I would stay away from install related stuff.

jeff-cohere replied:
Removed.

bartgol commented on components/scream/docs/build.md:
````diff
@@ -0,0 +1,93 @@
+Create a test directory in `RUN_ROOT_DIR` and use CMake to configure your build.
+This usually looks something like the following:
+```
+mkdir -p $RUN_ROOT_DIR/test
````
The test subdir will be empty, since there's no info in the cmake config line that tells cmake to build inside that folder. I would simply remove the 'test' folder from the guide.

jeff-cohere replied:
Removed.

bartgol commented on components/scream/docs/build.md:
````diff
@@ -0,0 +1,93 @@
+P3 regression tests (again, from the `RUN_ROOT_DIR/test` directory), use
+
+```
+ctest -R p3_regression
````
Let's also mention that we have labels. One can see the available labels with `ctest --print-labels` and then see what tests a label would run with `ctest -L my_label -N`.

jeff-cohere replied:
Done.

bartgol commented on components/scream/docs/build.md:
````diff
@@ -0,0 +1,104 @@
+To see which tests are associated with a given label (e.g. `p3_regression`), use
+
+```
+ctest -L p3_regression -N
````
Argh, I just noticed p3 tests do not set labels. As an example, we could use 'driver', since the AD does set a label ('driver') on tests.

Btw, we should encourage people to use labels, but we should probably establish a set of predefined labels, so that we don't end up with as many labels as tests. A topic for another day.

jeff-cohere replied:
Replaced with `driver`.

jeff-cohere commented:
Thanks for the feedback, guys.Okay, I think I've addressed your comments so far. @PeterCaldwell , I rewrote README.md. See what you think.

PeterCaldwell commented:
I just pushed a few tweaks to build.md that I couldn't bear to live without. I think it looks great and we should merge it now. This is a great example of a PR that should be merged without bothering with testing...

jeff-cohere commented:
Great. @bartgol , what's the next step? Can you do the magic merge-without-testing thing?

PeterCaldwell commented:
I can do it. I was just being lazy.
